### ISSUE TITLE
"can't set attribute" using nengo_ocl

### ISSUE THREAD

**Issue by bmorcos** [nengo/nengo#1180]
nengo_ocl up to date on `master`
pyopencl up to date on `master`
nengo up to date on `fixed-decoder-solver`

Getting the following:

```
AttributeError                            Traceback (most recent call last)
<ipython-input-4-4b371992e2da> in <module>()
      7     for j, D2 in enumerate(D2s):
      8         print(D1, D2)
----> 9         t = evaluate_time(N, D1, D2)
     10         data[i,j] = t
     11 

<ipython-input-3-e27e3c933222> in evaluate_time(N, D1, D2, run_time, dt)
      1 def evaluate_time(N, D1, D2, run_time=1.0, dt=0.001):
      2     m = make_model(N, D1, D2)
----> 3     sim = nengo_ocl.Simulator(m, dt=dt)
      4 
      5     start = timeit.default_timer()

/home/ben/.local/lib/python3.5/site-packages/nengo_ocl-1.0.1.dev0-py3.5.egg/nengo_ocl/simulator.py in __init__(self, network, dt, seed, model, context, n_prealloc_probes, profiling, if_python_code, planner)
    314         # --- create these first since they are used in __del__
    315         self.closed = False
--> 316         self.model = None
    317 
    318         # --- arguments/attributes

AttributeError: can't set attribute
```

Thanks @tbekolay for the workaround:

Add the following to the `Simulator` class in `nego/simulator.py`:

```
    @model.setter
    def model(self, newmodel):
        self._model = newmodel
```

**Comment by tbekolay** [comment]
This is because of #1173; we made the model private, but `nengo_ocl.Simulator` needs access to it. We'd essentially have to reverse that decision to avoid changing OCL (I'd like to avoid having to change OCL, since it would make compatibility trickier), though perhaps there's a clever way to make it "protected" instead of "private"?

**Comment by drasmuss** [comment]
I'd be fine with just rolling back #1173.  There was nothing super necessary about that change, and now we have a concrete example of why people might legitimately want access to `sim.model`.

**Comment by jgosmann** [comment]
I would like to challenge that the access to `sim.model` is completely legitimate. Given how often we (I?) accidentally break nengo_ocl, it might indicate that both simulator classes are too tightly coupled (especially as they are in two separate packages). Looking at `nengo_ocl.Simulator` I noticed that this class overwrites `__init__`, but does not call the super constructor (which is usually bad practice IMO). That way a portion of the `nengo.Simulator` code is not actually used and it obscures what can be changed in the `nengo.Simulator` without breaking `nengo_ocl.Simulator`.

I would argue that `nengo_ocl.Simulator` either should not inherit from `nengo.Simulator` and implement the remaining functions itself (after already abandoning the inherited constructor) or if there is legitimate shared code there should be a common abstract base class from which `nengo.Simulator` and `nengo_ocl.Simulator` inherit. In either case `nengo_ocl.Simulator` would define it's own `model` attribute with full control over it.

This would require some refactoring of course, so if we're pressed for time, reverting part (!) of #1173 would certainly be an option. But otherwise we should carefully think about whether the current implementation of `nengo_ocl.Simulator` is appropriate.

**Comment by tbekolay** [comment]
I agree, if we're expecting a certain API for every backend's `Simulator` class, an abstract base class might be in order. And since OCL uses Nengo's builder, it might be the case that we should be calling the superclass init... I'll take a look at this and propose something this week.

**Comment by tbekolay** [comment]
The abstract base class is something we should do in the long term, perhaps as part of #1189, but right now it's not feasible. As such, I've made https://github.com/nengo/nengo_ocl/pull/128 to fix this in the short term (without reverting #1173). I'll close this now as the issue will soon be fixed, and leave the bigger change to #1189.
